### PR TITLE
Add RandomBrightness preprocessing layer.

### PR DESCRIPTION
**Pull request by qlzh727**
Fix https://github.com/keras-team/keras-cv/issues/87.

As a demo (produced by the demo script).

Original image:
original

Augmented image (scale by -0.5 ~ 0.5
augmented
 randomly selected)

### CODE REVIEW

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+
+class RandomBrightness(tf.keras.layers.Layer):
+  """Randomly adjust brightness for the a RGB image.
+
+  This layer will randomly increase/reduce the brightness for the input RGB image.
+  During inference time, the output will be identical to input. Call the layer with
+  training=True to adjust brightness of the input.
+
+  Note that same brightness adjustment will be apply to all the images in the same
```
Comment (bhack) [inline keras_cv/layers/preprocessing/random_brightness.py:28]: Is this what we want?  Does this have any positive/negative significant impact on convergence?
https://github.com/tensorflow/models/blob/master/official/vision/beta/projects/simclr/dataloaders/preprocess_ops.py#L30

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      self._scale = [-scale, scale]
+    else:
+      raise ValueError(_SCALE_VALIDATION_ERROR +  f'Got {scale}')
+
+  @staticmethod
+  def _check_scale_range(input_number):
+    if input_number > 1.0 or input_number < -1.0:
+      raise ValueError(_SCALE_VALIDATION_ERROR +  f'Got {input_number}')
+
+  def call(self, image, training=None):
```
Comment (bhack) [inline keras_cv/layers/preprocessing/random_brightness.py:90]: Are we handling a single image or images/batch?

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  This layer will randomly increase/reduce the brightness for the input RGB image.
+  During inference time, the output will be identical to input. Call the layer with
+  training=True to adjust brightness of the input.
+
+  Note that same brightness adjustment will be apply to all the images in the same
+  batch.
+
+  Args:
+    scale: Float or a list/tuple of 2 floats between -1.0 and 1.0. The scale is
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/random_brightness.py:32]: should we expose image pixel boundaries as well?  I'm starting to think we should support all formats for pixel values (i..e [0,1], [-1, 1], [0, 255]).

This seems like a decision we should make and keep consistent across all of our layers

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_brightness.py:28]: great link - yeah, if possible we should avoid applying the same augmentations to an entire batch unless it is extremely difficult to avoid doing so.

**Comment by LukeWood** [comment]
Thanks Scott for the feature add!

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 1 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/random_brightness.py:28]: Also if it is the same on the entire batch why we are not just layer wrapping:
https://www.tensorflow.org/api_docs/python/tf/image/random_brightness

Comment (bhack) [reply keras_cv/layers/preprocessing/random_brightness.py:28]: Also as we have discussed in https://github.com/keras-team/keras-cv/pull/91#issuecomment-1027846627 we still need to evaluate if we want to expose the op in the public API, keep it only in the keras-cv private API or just embedded in the layer itself.

In the case we want to expose the op in the public API are we going to almost duplicate `tf.image.random_brightness`?

Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: Updated. The new implementation will generate different adjustment for each of the image in the batch.

On a side note, tf.image.random_brightness has some assumption about the dtype and the range of the value. Eg when dtype is int, it expect the value to range from [0, 255]. When the dtype is float, it will assume the value is ranged [0.0, 1.0]. This means if you give it a float32 image with value [0, 255], then it will almost adjust nothing, since the delta is only [0, 1] and not scaled with * 255. 

The current implementation in this PR is only expecting RGB colored image, and all the dtypes should work.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 2 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: We handle both single image and batch image. See the docstring and unit test.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 3 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: For brightness/contrast and any color related augmentation, I feel we should stay with RGB color (0~255). The normalization of the input (for model) should happen after the color adjustment.

**Comment by qlzh727** [comment]
It seems that the test for random_cutout are failing. We should fix or disable it for now.

**Comment by LukeWood** [comment]
> It seems that the test for random_cutout are failing. We should fix or disable it for now.

Weird - I just ran locally and it worked.  Let me look into what's going on here.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 2 above)
Comment (bhack) [reply keras_cv/layers/preprocessing/random_brightness.py]: Yes I know but I meant that we could standardize the names in our API.
Sometimes we use image in other ops images when we handle batches.

Comment (chjort) [reply keras_cv/layers/preprocessing/random_brightness.py]: I think `inputs` would be appropriate. This follows what every other keras layer calls it.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 3 above)
Comment (innat) [reply keras_cv/layers/preprocessing/random_brightness.py]: I also think so, like there is a separate [rescaling layer](https://keras.io/api/layers/preprocessing_layers/image_preprocessing/rescaling/) that can be used for pixel normalization after all transformation.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 2 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: Done. changed to  `inputs` for consistency.

Comment (bhack) [reply keras_cv/layers/preprocessing/random_brightness.py]: Ok but, also if minor, we have the same issue in the next internal/private method. I think we need to make some final choice.

File: keras_cv/layers/preprocessing/random_brightness.py (hunk 1 above)
Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: I think the difference between the current PR and tf ops is quite large and have different assumption

1. Scale the delta based on the input dtype (this is the major one, and has been apply to all the tf.image ops, which I don't think can be easily changed)
2. Support of batch inputs. All the existing tf.image ops only take care of single image input. Updating all of them to have batch input support might require more work (but doable).
3. Output value clip. (Minor and doable). 
4. Uneven delta range. (Minor and doable). 

I am not sure who is the active API owner of tf.image ops, @fchollet for more inputs.

Comment (bhack) [reply keras_cv/layers/preprocessing/random_brightness.py]: 2. [Are you sure](https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/ops/image_ops_impl.py#L1958)?

In the end probably it is better to take the ownership of `tf.image` here and handle the deprecation warning PRs in TF instead of growing incompatible API between TFA/TF/Keras-*.

Then I understand the everyone care only about its ownership but I think that we could improve the library perception to the users/contributors if we think as an ecosystem.

Comment (qlzh727) [reply keras_cv/layers/preprocessing/random_brightness.py]: Oh, you are right, my mistake. The current PR apply different adjustment to each image, but the tf.image apply the same adjustment.

Checked with @fchollet  offline. It seems that there isn't an active owner for tf.image at this moment. We will have some discussion about how we are going to move forward.

**Comment by bhack** [comment]
@mattdangerw Also internally `tf.keras.layers.RandomConstrast` implementaton relies on the `tf.image` ops:

https://github.com/keras-team/keras/blob/v2.8.0/keras/layers/preprocessing/image_preprocessing.py#L1183

Here in the brightness the ops is forked with a slightly different behaviour.

**Comment by LukeWood** [comment]
> A couple comments on consistency within Keras. It seems like bad UX that:
> 
> * Random contrast will live and be documented in core Keras, as `tf.keras.layers.RandomConstrast`.
> * Random brightness live will and be documented here as `keras_cv.layers.RandomBrightness`.
> * The two layers will have different expectations around input ranges.
> 
> I am not super opinionated on a solution. But I think we should:
> 
> * Consider landing this in core keras, or aliasing RandomContrast here, so the two can be used in the same way.
> * If we stick with the 0, 255 range here, consider doing the same for RandomContrast, if no one is relying on the current behavior.

I like the idea of aliasing RandomContrast.  It will quickly become an outlier as more and more image based preprocessing layers become hosted in KerasCV.  Changing the old behavior to expect [0, 255] seems reasonable if KPLs have not been around long enough that this breaking change will be too difficult to get submitted.

**Comment by mattdangerw** [comment]
> I like the idea of aliasing RandomContrast. It will quickly become an outlier as more and more image based preprocessing layers become hosted in KerasCV. Changing the old behavior to expect [0, 255] seems reasonable if KPLs have not been around long enough that this breaking change will be too difficult to get submitted.

Unless we do some sort of bulk solution, I think that tension will continue to persist somewhat given that there are [10ish](https://keras.io/guides/preprocessing_layers/#available-preprocessing) image preprocessing layers in core Keras.

I would actually vote for `RandomBrightness` in core keras. That would give us an informal split of the "really simple" image augmentation (flips, rotations, brightness, contrast) living in core, and the more complex (cutmix, cutout, solarization, histogram equalization, randaugment) living in keras-cv. I would find that somewhat intuitive, and we can (and probably have to) cross link from one to the other in our documentation.

**Comment by LukeWood** [comment]
> > I like the idea of aliasing RandomContrast. It will quickly become an outlier as more and more image based preprocessing layers become hosted in KerasCV. Changing the old behavior to expect [0, 255] seems reasonable if KPLs have not been around long enough that this breaking change will be too difficult to get submitted.
> 
> Unless we do some sort of bulk solution, I think that tension will continue to persist somewhat given that there are [10ish](https://keras.io/guides/preprocessing_layers/#available-preprocessing) image preprocessing layers in core Keras.
> 
> I would actually vote for `RandomBrightness` in core keras. That would give us an informal split of the "really simple" image augmentation (flips, rotations, brightness, contrast) living in core, and the more complex (cutmix, cutout, solarization, histogram equalization, randaugment) living in keras-cv. I would find that somewhat intuitive, and we can (and probably have to) cross link from one to the other in our documentation.

yeah, that's a really good point.  I underestimated how many image augmentations were in core Keras.  RandomBrightness does seem to fit the mould a bit better in core Keras.  I'm on board to add it there.

**Comment by bhack** [comment]
Honestly we are going to strongly confuse users on the namespace to use or the wheel to install to find a specific CV preprocessing layer.
I suppose that we are going to confuse contributors on what repo  they are going to propose a PR/ticket for a CV preprocessing ops.

How solarize is different from brightness? It is hard to semantically grasp this namespace difference as the TF user base has both in the same `tf.image` namespace:

https://www.tensorflow.org/api_docs/python/tf/image

**Comment by qlzh727** [comment]
> > > I like the idea of aliasing RandomContrast. It will quickly become an outlier as more and more image based preprocessing layers become hosted in KerasCV. Changing the old behavior to expect [0, 255] seems reasonable if KPLs have not been around long enough that this breaking change will be too difficult to get submitted.
> > 
> > 
> > Unless we do some sort of bulk solution, I think that tension will continue to persist somewhat given that there are [10ish](https://keras.io/guides/preprocessing_layers/#available-preprocessing) image preprocessing layers in core Keras.
> > I would actually vote for `RandomBrightness` in core keras. That would give us an informal split of the "really simple" image augmentation (flips, rotations, brightness, contrast) living in core, and the more complex (cutmix, cutout, solarization, histogram equalization, randaugment) living in keras-cv. I would find that somewhat intuitive, and we can (and probably have to) cross link from one to the other in our documentation.
> 
> yeah, that's a really good point. I underestimated how many image augmentations were in core Keras. RandomBrightness does seem to fit the mould a bit better in core Keras.
> 
> Ok, I'm on board to add it there.

I also like the idea of having core image KPL in keras core, and have extended KPL in keras-cv, in the meantime, the core image KPL should probably be aliasing here. 

For the behavior difference of [0, 255] vs [0, 1] in the existing keras image KPL, we should probably fix the discrepancy and unify to [0, 255]. This means we will update the image KPL in core keras.

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+_SCALE_VALIDATION_ERROR = (
+    "The `scale` should be number or a list of two numbers "
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/random_brightness.py:18]: "The `scale` argument should be a number (or a list of two numbers) "
"in the range [-1.0, 1.0]. "

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    This layer will randomly increase/reduce the brightness for the input RGB image.
+    During inference time, the output will be identical to input. Call the layer with
+    training=True to adjust brightness of the input.
+
+    Note that different brightness adjustment will be apply to each the images in the
+    batch.
+
+    Args:
+        scale: Float or a list/tuple of 2 floats between -1.0 and 1.0. The scale is
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/random_brightness.py:34]: For naming consistency with other preprocessing layers, we should use `factor`, see e.g. https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_rotation/

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+            used to determine the lower bound and upper bound of the brightness
+            adjustment. A float value will be choose randomly between the limits.
+            When -1 is chosen, the output image will be black, and when 1.0 is
+            chosen, the image will be fully white. When only one float is provided,
+            eg, 0.2, then -0.2 will be used for lower bound and 0.2 will be used for
+             upper bound.
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/random_brightness.py:40]: Should we also support integers? Interpreted as a range in absolute levels, e.g. `32` would be interpreted as a +-32 shift of the channels

File: keras_cv/layers/preprocessing/random_brightness.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            self._scale = [-scale, scale]
+        else:
+            raise ValueError(_SCALE_VALIDATION_ERROR + f"Got {scale}")
+
+    @staticmethod
+    def _check_scale_range(input_number):
+        if input_number > 1.0 or input_number < -1.0:
+            raise ValueError(_SCALE_VALIDATION_ERROR + f"Got {input_number}")
+
+    def call(self, inputs, training=None):
```
Comment (fchollet) [inline keras_cv/layers/preprocessing/random_brightness.py:93]: Augmentation preprocessing layers should default to `training=True` (default should be to apply the transformation). You can assume that `training` is a Python bool in what follows.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by qlzh727** [comment]
Closing this PR now. Since we have keras.layers.RandomContrast, Keras team decided to keep the RandomBrightness in the core keras as well since both of them are very close to each other. See https://github.com/keras-team/keras/blob/9628af85a0a2cb04cf433b1ad991017b70ae2005/keras/layers/preprocessing/image_preprocessing.py#L1393 for more details.
